### DISCUSSION TITLE
Add a Spreadsheet Layout

### DISCUSSION THREAD

**Discussion by benhaynes** [directus/directus#3874] [Feature Requests]
This would be a layout similar to the table layout, but would allow editing fields directly on the page. I'm torn between two options:

1. Spreadsheet style, showing and editing raw values
2. Table Layout Plus... where clicking a field will open a context popup with the proper interface

Here's a quick mockup I made...

![Item-Browse-Spreadsheet](https://user-images.githubusercontent.com/522079/106406952-dcb07580-6408-11eb-9063-0727814b0b3b.png)

**Comment by cynova** [comment]
Another amazing idea!!!
You're trying to make every last other CMS obsolete, aren't you? :P

I'd prefer a third option: Use the correct interfaces and only make them pop up if necessary from a layout perspective. That way in the mockup above probably only the Image fields would need context popup, as text/dropdown should work with really little space.

Of course the decision which mode to use could be manual (but per field), in the layout settings.

**Reply by benhaynes** [reply]
Hahah, thank you!

Regarding "Option 3", it might seem like most interfaces can fit in that small space, but most interfaces are actually a bit too large. IMO, a good alternate option would be for the Layout itself to have dedicated ways of managing each datatype, and perhaps a "right-click the cell" menu that has an option for editing with the full interface in a drawer. This might be the best of both worlds.

For example, here is how we might edit values (with right-click always available to use the native interface):

* String/Text: Raw
* Int/Number: Raw
* JSON/CSV/UUID/Hash: Raw
* Date/Time/etc: Raw, "Select" icon (dropdown)
* Boolean: Built-in cell toggle
* O2M/M2M/M2A/Files: Item Count (drawer), "New" icon (drawer), "Select" icon (drawer)
* Translation: Translation Count (drawer), "New" icon
* Image/File: Thumbnail (drawer), "Deselect" icon, and "Select" icon
* M2O/Dropdowns: Raw, "Select" icon (built-in cell dropdown), "Deselect" icon

**Comment by soyjimmysaenz** [comment]
This would be a killer feature definitely! Currently we are trying to migrate a system from google sheets to Directus, and a lot of people are reluctant because they were accustomed to the spreadsheet format.

**Comment by eliteproxy7** [comment]
@benhaynes 
There was a feature in ZOHO CRM where it would allow you to edit your view in "google sheets" style view and then "sync" it back. This even added new records where the ID was left blank. I don't think it was using "actual" google sheets but perhaps a component. This could also help with the "csv import" ui?

**Reply by benhaynes** [reply]
Interesting! I think the Spreadsheet Layout would be great for quick/bulk edits, but we can still have "Import" (eg: CSV) as a separate feature, so it works for _any_ Layout. That said, there's plenty of room for discussion with how this will work. 👍

**Reply by Jeyakumarrathnaselvi** [reply]
Like to sponsor the activity within my limit if anyone can take it up.

**Comment by johannesmutter** [comment]
An editable spreadsheet view would be fabulous. I’d prefer the option to specify for each column wether a single click goes into RAW editing or wether a popup should be displayed. I think Airtable is a good reference.

I’m currently using NocoDB just for this feature. My primary use case is copy & paste data from sheets/ CSV and bulk editing.

**Comment by cdwmhcc** [comment]
Great!

**Comment by loongmxbt** [comment]
Hi, recently we've made a datagrid interface for non-relational tables. It can store both json and csv table data, supports bulk editing, chart displaying. Json data supports both records-like data `[{col1: v1, col2: v2, col3: v3}, {}, {}]` and list-like data `{ 'col1': [v11, v12, v13], 'col2': [v21, v22, v23] }`.
https://github.com/seymoe/directus-extension-vgrid-interface
<img width="829" alt="image" src="https://user-images.githubusercontent.com/5986600/165288760-ac1cb8d1-d24f-42bb-a547-e51f578b2b2f.png">

**Reply by benhaynes** [reply]
Ohh, I like this... nice job!

**Comment by julbd** [comment]
This is an awesome feature ! ⭐️

I'd advice you not to build the spreadsheet yourself but to reuse a dedicated component (like [Revogrid](https://github.com/revolist/revogrid), with proper patches and customization ?) or you'll end up with endless requests for new features everyone expect from spreadsheets but were not in your 80/20 plan (selection, copy paste, moving values, infinite loading...).

You can also look how revogrid manage the "edition interfaces" you discussed above : cells displays formatted values (like the classic directus view), until you double-click to edit them and they display raw values with sometimes buttons and popup.

For directus, as a "quick first step", you could simply reuse the edition _interfaces_ to put in place of the cells (or in top of, if too large, with a shadow effect) when (double-)clicked, with auto focus & popup opened. This would not be visually perfect for every data types, but this would enable you to reuse all _displays_ and _interfaces_ as is, without the drawer arduousness.

English is not my first langage and i'm not confident with directus internals, so here is full a summary of my proposal 😄 :
- External component (revogrid?) to handle the spreadsheet structure & associated features.
- In view mode, use _displays_ in cells.
- In edition mode (cell click or double-click), show _interfaces_ within the cell, with focus and popup. If the interface is larger than the cell (detected on runtime), display it on top of the cell with a shadow.
- For maximum compatibility, interfaces could declare their "in-spreadsheet" mode : `cell` (default) or `drawer` (opens drawer on click).
- You could display a `...` menu button on hover inside the cell (view mode) or next to the interface (edition-mode), with _Open in drawer_ option beside the others field options (_Edit raw value_, _Copy raw value_, _Erase value_...).

**Reply by katywings** [reply]
Great suggestion 👍, on the: `you could simply reuse the edition interfaces to put in place of the cells`, i am wondering if this is possible in a custom layout plugin, then we could make a prototype via a custom layout before doing everything in the core 🤔

**Reply by benhaynes** [reply]
Really great points. And I agree, @julbd, we don't want to build a custom spreadsheet/formula library... that's just asking for trouble. That said, I think there actually might be two (or three!) different Layouts here... let me outline:

## 1. Table Layout
**Difficulty: Normal**
We already have a table layout that is great and pretty full-featured, but it's lacking a few things that would get it closer to this spreadsheet idea. This would allow for quick edits, still uses the nice interfaces, etc.
* The ability to toggle the entire view to show **raw** database values instead of the Display's presentation
* "Inline" editing. I put that in air quotes because I think we might want to have that access behind a click. So you (right-)click the cell (when looking at it in presentation or raw mode) and you get an option to quickly edit the value in an inline popover that shows the configured interface.

## 2. Spreadsheet Panel
**Difficulty: Difficult or Normal**
The main difference here is that I think we want to lean into the idea of actual formulas... which would be quite different from the idea of mapping/showing fields, values, displays, and interfaces. This honestly seems more like a readonly (in terms of the database, not the GUI) sandbox, where you can create dynamic presentations using the database content. If I'm on the right track, this actually might be better within Insights... which is more about customizable presentation/visualization, and less about viewing and managing exact fields.

## 3. Database Administration Section
**Difficulty: Normal or Easy??**
My thinking here is to include a more technical tool for admins to efficiently view/manage database content. No displays, no interfaces, no formulas... just what you'd expect to see within a database administration tool/client like phpMyAdmin (gasp!), Sequel Pro, Table Plus, etc. I think this should be _admin only_ because it's more technical (less intuitive, especially for FKs), and might be better without access control restrictions. We'd still need to think through validation (should be fine to be async) and revisions... would every cell change have activity+revision data saved? 🤔 

Maybe this last option isn't a Layout, but a dedicated section with Settings (since that's exclusive to admins). Either way, I think this would be really nice to compete against other backend-as-a-service developer tooling.

----

I'd love to hear what the community thinks of this more scoped approach! 📢 

> i am wondering if this is possible in a custom layout plugin, then we could make a prototype via a custom layout before doing everything in the core

Good question for @rijkvanzanten ... but I think 1+3 above should be in Core, so we could prototype on PRs.

**Reply by eliteproxy7** [reply]
The reason I think the Google sheets type of integration is better vs a custom table view with editing capability, this gives you full spreadsheet usability to manage the data for import / bulk editing. 

Zoho opes up a new tab with the spread sheet with a save button once saved it syncs. It basically does an import (we can use our current import API)

the formulas are not saved but the values is what gets updated. 

for Postegres users there are allot of Postegres & google sheets integration, this may be a workaround for now. Just integrate your collection with the DB so your updating the database directly. (Zappier has a zap for this)

**Reply by benhaynes** [reply]
@eliteproxy7 — what you're describing seems separate, though. That seems more like a new abstracted datastore adapter that allows pulling in Google Sheets as an external data source, pulling the title as the collection, and mapping columns/rows to fields and a PK. That would mean you could view/manage Google Sheets in any Directus views... different layouts, etc.

**Reply by julbd** [reply]
> So you (right-)click the cell

The right-click is not widely used elsewhere in directus (or not used at all) and would not be expected here. I would rather recommend a simple click or a discreet `...` button if you need more options.

> Spreadsheet Panel

You could reach the same result with a simple "Table Layout" + advanced [computed fields](https://github.com/directus/directus/discussions) with functions. In this case, these advanced computed fields would have dedicated display and edition interface, like other fields types, and no spreadsheet-specific code (and can easily be used anywhere) !

And for presentation/visualization, the Insights panel already allows to display collections. I don't think about specific needs that would not be filled by collection previews with table layout in Insights.

80/20 + maximum re-usability 😛

> Database Administration Section

Personally, this is not something I expect from Directus. It provides me a very good data management interface on top of my database, I almost never feel the need to break the abstraction, and when I need to (repetition), I'll find the proper and dedicated tools for that. I think Directus should leave that part of the job to others (all hosts already provides a database management interface).

**Reply by benhaynes** [reply]
> The right-click is not widely used elsewhere in directus (or not used at all) and would not be expected here. I would rather recommend a simple click or a discreet ... button if you need more options.

We're looking to change that, but want to do it "in a big push" so it's intentional, consistent, and documented. As for just being "on click", that is the event to open the item detail page, so you'd have to enter some global context for an "editing mode" first.

> You could reach the same result with a simple "Table Layout" + advanced [computed fields](https://github.com/directus/directus/discussions) with functions. In this case, these advanced computed fields would have dedicated display and edition interface, like other fields types, and no spreadsheet-specific code (and can easily be used anywhere) !

Very true! I think the main thing to think through here is how can any user create and configure these alias/computed fields on the fly within the Layout itself. Does that affect API output, etc? Assuming this would be possible by admins and normal users alike, we'd need a way to confgure if this is allowed by non-admins. This is why this is the most difficult direction.

> And for presentation/visualization, the Insights panel already allows to display collections. I don't think about specific needs that would not be filled by collection previews with table layout in Insights.

True, if any Collection/Layout can be added to Insights, then this is moot.

> Personally, this is not something I expect from Directus. It provides me a very good data management interface on top of my database, I almost never feel the need to break the abstraction, and when I need to (repetition), I'll find the proper and dedicated tools for that. I think Directus should leave that part of the job to others (all hosts already provides a database management interface).

Not sure if I'm 100% onboard with this one. To be clear, I'm not really talking about the schema/DDL side of things... just a fast, raw way to see the database as-is (no presentation). It also is a quick way to remove all the "icing on the cake" that Directus provides (alias fields, displays, relational traversal, etc) so you can audit the actual DB values.

I do agree that these tools are readily available and they're going to be better (more robust) at handling this. But I don't see a reason why this can't be included as an admin feature if it's not a very heavy engineering lift or bloated code. Anywho, always worth weighing the pros/cons!

**Reply by julbd** [reply]
> in a big push

Ok, good idea !

> Does that affect API output, etc? Assuming this would be possible by admins and normal users alike, we'd need a way to confgure if this is allowed by non-admins. This is why this is the most difficult direction.

I see. The need would not only to have "dynamic fields", but to let all users create the dynamic fields they need in a shared environment. Well, in classic shared spreadsheets, all fields are shared anyway, so you can start by enabling all users to create _alias_ fields on the fly (because they don't change the database structure). And you'll see if there is requests for more advanced features (like user/group-specific fields).

**Reply by julbd** [reply]
> just a fast, raw way to see the database as-is (no presentation). It also is a quick way to remove all the "icing on the cake" that Directus provides (alias fields, displays, relational traversal, etc) so you can audit the actual DB values.

As far as it understand, this meet the other feature you were describing : "the ability to toggle the entire view to show raw database values instead of the display's presentation". So maybe "Raw data" toggle in the collection header that switch to a preset with :
- Table view
- All fields displayed ?
- Raw database values.
- No alias fields or with a debug value.

**Reply by benhaynes** [reply]
> As far as it understand, this meet the other feature you were describing : "the ability to toggle the entire view to show raw database values instead of the display's presentation".

Yeah, that gets you most of the way. To get those four items is a lot of extra config on the no-code side of things (the layout)... but there is a good amount of overlap.

**Comment by boo-code** [comment]
Maybe https://nocodb.com can be used as a spreadsheet for the collections? Directus can integrate NocoDB by sharing the same DB connection, and can have a button "Open in NocoDB" (opens collection in a new tab and in NocoDB loaded) in the collection toolbar.

**Reply by benhaynes** [reply]
Interesting. This platform actually seems far more robust than a library we can leverage... more like a secondary competitor. Still, we'll look into it!

**Comment by taviks** [comment]
This is the primary thing preventing me from wanting to use 99% of all content management systems. 

So much of Directus is _incredible_, but without this feature I'm still reluctant to use it.

If Directus can ship this and execute it well it will be a true game changer (at least for me).

**Reply by cdwmhcc** [reply]
A save button is needed, accidents can happen if fields are accidentally modified.

**Reply by Jambris** [reply]
> A save button is needed, accidents can happen if fields are accidentally modified.

This is the risk you take when you give people access which is why you can filter biased on roles. 
A save button shouldn't be needed, or at the very least be optional as low priority.
This reduces the need for external spreadsheets which can allow for better data entry, which is a huge plus.
I wish i had the time to implement this myself :(

**Comment by martindemello** [comment]
the sadly discontinued https://github.com/basetool-io/basetool seems like a perfect fit for the missing piece; it is better than nocodb because it supports collaborative editing. not sure how hard it would be to integrate it with directus

**Comment by arladmin** [comment]
Did anyone ever manage to implement this?

Is there any plan, by Directus, to bring this feature to life?

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by zhengpd** [reply]
Hi @rijkvanzanten , has this feature been reviewed?

**Reply by rijkvanzanten** [reply]
Nope! Well I mean it's an idea coming from our own team, but we still gotta update the original posting to have the full RFC format so we can properly implement it 🙂 There's still a couple of unresolved questions we need to address, like 

- is this showing/editing exclusively raw database values in cells of a spreadsheet?
- is it more like inline editing in a table where it renders an interface in a popup?
- should it use displays to render the visible value, and switch to raw values in the cell on edit?

so yeah that should be reviewed and thought through still!

**Comment by arladmin** [comment]
@rijkvanzanten 


> Nope! Well I mean it's an idea coming from our own team, but we still gotta update the original posting to have the full RFC format so we can properly implement it 🙂 There's still a couple of unresolved questions we need to address, like
> 
> * is this showing/editing exclusively raw database values in cells of a spreadsheet?
> * is it more like inline editing in a table where it renders an interface in a popup?
> * should it use displays to render the visible value, and switch to raw values in the cell on edit?
> 
> so yeah that should be reviewed and thought through still!

My suggestion:

Have 2 modes for the table as a whole: default & raw.
- default : "inline editing in a table where it renders an interface in a popup" & "use displays to render the visible value"
- raw : "showing/editing exclusively raw database values in cells of a spreadsheet"

**Comment by julbd** [comment]
> - is this showing/editing exclusively raw database values in cells of a spreadsheet?
> - is it more like inline editing in a table where it renders an interface in a popup?
> - should it use displays to render the visible value, and switch to raw values in the cell on edit?

I would suggest that it depends on the display and interface :
- View : if the display supports "inline", use it in place, otherwise show raw data.
- Edit (on click): if the interface supports "inline", use it in place, otherwise show the interface in a popup.

The display and interface can have their behavior or style customized with a parameter indicating whether it is invocated in inline or classic context

**Comment by nihilo666** [comment]
Any news about the process for table / Bulk edit like a google sheet? Would be an amazing feature - I mean - actually it should be an essential thing.

**Comment by arladmin** [comment]
For Directus Studio to be usable as the primary tool (for various use-cases), this sort of functionality in the UX/UI becomes even more critical!

Otherwise, most of the use-cases for Directus Studio would continue to be:
- being relegated to either a sophisticated Database browser (like DBeaver, Beekeeper Studio etc.) 
- basic Internal apps with very primitive UX.

**Comment by rafarc21** [comment]
Any updates here? Really looking forward to this one!

**Comment by nachumie** [comment]
Would be great!!!

**Comment by phazonoverload** [comment]
We've just released a spreadsheet layout as a Directus Labs project factoring in many of the comments made here. 

- Fields can be edited directly from the spreadsheet, changing between displays and interfaces.
  - Types: string, text, bigInteger, integer, float, decimal, boolean, dateTime, date, time, timestamp.
  - Interfaces: boolean, datetime, input, input-autocomplete-api, select-color, select-dropdown, select-icon, slider.
- Either automatically save on field defocus, or make many changes and save them all at once.

It's available now in every Directus project via the Marketplace. Enjoy! 

![](https://github.com/directus/directus/assets/1461554/997a3f10-29c3-4f7a-880a-3d3296fb9d3e)

**Reply by katywings** [reply]
Thank you for the hard work 🥳👏🎉
